### PR TITLE
Remove duplicate drawing code from several track elements

### PR DESCRIPTION
**Pull request by karst**
There's a bunch of duplicate code that could be unified. Here I am picking out 0-25 25 and 25-0 as well as 0 with chainlift.
I found a few that could be written shorter as well, but I'll start with these.

### CODE REVIEW

File: src/openrct2/ride/coaster/LoopingRollerCoaster.cpp [hunk 1]
```diff
@@ -31,33 +31,9 @@ static void LoopingRCTrackFlat(
...
-                    session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15009), { 0, 0, height }, { 32, 20, 3 },
-                    { 0, 6, height });
-                break;
-        }
-        if (TrackPaintUtilShouldPaintSupports(session.MapPosition))
-        {
-            MetalASupportsPaintSetup(session, METAL_SUPPORTS_TUBES, 4, 0, height, session.TrackColours[SCHEME_SUPPORTS]);
-        }
+        PaintAddImageAsParentRotated(
+            session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15006 + direction), { 0, 0, height },
```
Comment (frutiemax) [inline src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:35]: Use the call with BoundBoxXYZ

File: src/openrct2/ride/coaster/LoopingRollerCoaster.cpp [hunk 2]
```diff
@@ -125,64 +103,22 @@ static void LoopingRCTrack25DegUp(
...
-                    session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15063), { 0, 0, height }, { 32, 20, 3 },
-                    { 0, 6, height });
-                break;
-        }
-        if (TrackPaintUtilShouldPaintSupports(session.MapPosition))
-        {
-            MetalASupportsPaintSetup(session, METAL_SUPPORTS_TUBES, 4, 8, height, session.TrackColours[SCHEME_SUPPORTS]);
-        }
+        PaintAddImageAsParentRotated(
+            session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15060 + direction), { 0, 0, height },
```
Comment (frutiemax) [inline src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:107]: BoundBoxXYZ

File: src/openrct2/ride/coaster/LoopingRollerCoaster.cpp [hunk 3]
```diff
@@ -125,64 +103,22 @@ static void LoopingRCTrack25DegUp(
...
-                    session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15035), { 0, 0, height }, { 32, 20, 3 },
-                    { 0, 6, height });
-                break;
-        }
-        if (TrackPaintUtilShouldPaintSupports(session.MapPosition))
-        {
-            MetalASupportsPaintSetup(session, METAL_SUPPORTS_TUBES, 4, 8, height, session.TrackColours[SCHEME_SUPPORTS]);
-        }
+        PaintAddImageAsParentRotated(
+            session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15032 + direction), { 0, 0, height },
```
Comment (frutiemax) [inline src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:113]: BoundBoxXYZ

File: src/openrct2/ride/coaster/LoopingRollerCoaster.cpp [hunk 4]
```diff
@@ -281,64 +215,22 @@ static void LoopingRCTrackFlatTo25DegUp(
...
-                    session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15055), { 0, 0, height }, { 32, 20, 3 },
-                    { 0, 6, height });
-                break;
-        }
-        if (TrackPaintUtilShouldPaintSupports(session.MapPosition))
-        {
-            MetalASupportsPaintSetup(session, METAL_SUPPORTS_TUBES, 4, 3, height, session.TrackColours[SCHEME_SUPPORTS]);
-        }
+        PaintAddImageAsParentRotated(
+            session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15052 + direction), { 0, 0, height },
```
Comment (frutiemax) [inline src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:219]: BoundBoxXYZ

File: src/openrct2/ride/coaster/LoopingRollerCoaster.cpp [hunk 5]
```diff
@@ -281,64 +215,22 @@ static void LoopingRCTrackFlatTo25DegUp(
...
-                    session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15027), { 0, 0, height }, { 32, 20, 3 },
-                    { 0, 6, height });
-                break;
-        }
-        if (TrackPaintUtilShouldPaintSupports(session.MapPosition))
-        {
-            MetalASupportsPaintSetup(session, METAL_SUPPORTS_TUBES, 4, 3, height, session.TrackColours[SCHEME_SUPPORTS]);
-        }
+        PaintAddImageAsParentRotated(
+            session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15024 + direction), { 0, 0, height },
```
Comment (frutiemax) [inline src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:225]: BoundBoxXYZ

File: src/openrct2/ride/coaster/LoopingRollerCoaster.cpp [hunk 6]
```diff
@@ -539,64 +427,22 @@ static void LoopingRCTrack25DegUpToFlat(
...
-                    session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15059), { 0, 0, height }, { 32, 20, 3 },
-                    { 0, 6, height });
-                break;
-        }
-        if (TrackPaintUtilShouldPaintSupports(session.MapPosition))
-        {
-            MetalASupportsPaintSetup(session, METAL_SUPPORTS_TUBES, 4, 6, height, session.TrackColours[SCHEME_SUPPORTS]);
-        }
+        PaintAddImageAsParentRotated(
+            session, direction, session.TrackColours[SCHEME_TRACK].WithIndex(15056 + direction), { 0, 0, height },
```
Comment (frutiemax) [inline src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:431]: BoundBoxXYZ

**Review by frutiemax — COMMENTED** [review]
You should use the call with BoundBoxXYZ instead of the one with the separate CoordXYZ

File: src/openrct2/ride/coaster/LoopingRollerCoaster.cpp (hunk 1 above)
Comment (karst) [reply src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:35]: Could you give me an example for this for reference?

Comment (frutiemax) [reply src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:35]: Look at PaintAddImageAsParent, there is a function overload that uses BoundBoxXYZ.

Comment (karst) [reply src/openrct2/ride/coaster/LoopingRollerCoaster.cpp:35]: As mentioned in my other PR, I think it's a bit beyond the scope of this specific PR. Would be better for a separate.

**Comment by duncanspumpkin** [comment]
I completely forgot about this PR when looking at https://github.com/OpenRCT2/OpenRCT2/pull/19390 its going to have to be redone.

**Comment by karst** [comment]
No worries, I'll look into it sometime again 🐱

**Comment by ZehMatt** [comment]
@karst could you try to get this working again?

**Comment by karst** [comment]
Biggest annoyance when a merge/revert causes lint commit message to fail.

**Comment by karst** [comment]
I believe I have had all ride types that use this logic. Sadly G1 is an absolute mess for some rides, so I can't write a simple algorithm to select those (with simple I mean literally 0 to 3 for sprite selection).

Please let me know if I missed any rides, or if I messed something up somewhere. I very much got in a flow so I haven't exactly playtested it.

This should take care of most flat, gentle, flat to gentle and gentle to flat pieces.

File: src/openrct2/ride/coaster/BolligerMabillardTrack.hpp [hunk 7]
```diff
@@ -39,39 +39,22 @@ static void BolligerMabillardTrackFlat(
 {
     if (trackElement.HasChain())
     {
-        static constexpr const uint32_t imageIds[] = {
```
Comment (Gymnasiast) [inline src/openrct2/ride/coaster/BolligerMabillardTrack.hpp:42]: I’m not quite sure about removing this - I thought that @duncanspumpkin wanted to keep arrays or even introduce more sophisticated classes for this.

Comment (karst) [reply src/openrct2/ride/coaster/BolligerMabillardTrack.hpp:42]: I would like to make a noted document somewhere with all sprites. Especially with notes if their directions aren't in the same system from 0 to 3, which I have sadly come across too often today. But that way I hope all track pieces can be united on the same code. I know that TrackPaint.cpp is supposed to that, but I genuinely cannot make cake of that file. Might have to rewrite it in a system like this. I'll post the ideas to the track type discussion.

**Comment by karst** [comment]
Well I cant seem to figure out how to rename commits [ef8eaa5](https://github.com/OpenRCT2/OpenRCT2/pull/19286/commits/ef8eaa537c94cd4507e4fc28b0bcbce7ca330fe6) to [4dcb30e](https://github.com/OpenRCT2/OpenRCT2/pull/19286/commits/4dcb30ec9e177839855593d9cc52df538f058c62) as they are too long for Clang to enjoy it.

**Comment by Broxzier** [comment]
How did they end up as merge commits? Ideally, the commits from those merged branches are present as commits of this PR's branch.

**Pull request closed without merging**